### DISCUSSION TITLE
Splitting a schema across multiple sections of a documentation site

### DISCUSSION THREAD

**Discussion by gnidan** [jy95/docusaurus-json-schema-plugin#223] [Ideas]
I'm looking to author a Docusaurus site for a collection of several complex JSON Schemas. These schemas are recursive, but also just too big to explain properly in a single collapsible `<JSONSchemaViewer />`. I want to be able to document my schema in parts, with distinct pages or sections that contain individual `<JSONSchemaViewer />`s for individual definitions.

This plugin already allows specifying a JSON Pointer (which is essential!), but I'd also like to be specify known permalinks for subschemas when creating each `<JSONSchemaViewer />`. I'm imagining something like this:

<details>
<summary>Example interface to tell JSONSchemaViewer about known permalinks</summary>

```tsx
<JSONSchemaViewer
  {...otherOptions}
  knownRefs: {
    "#/$defs/ComplexStandaloneSubschemaWithSeparateDocumentation": "/spec/concepts/#documentation-for-complex-standalone-subschema"
  }
  />
```

</details>

In this example, I'm imagining this plugin to keep track of `"$ref"`s and render a component that hyperlinks to the corresponding page instead of recursively rendering the schema.[^ignoring-equivalent-ids]

Incidentally, this seems to be a viable way to workaround the lack of support for recursive `$ref`s. I've confirmed this already in my initial hacking, but I'm doing it with a wrapper around `JSONSchemaViewer` that _preprocesses_ my schema object by finding/rewriting all `{ $ref: ..., ...others }` objects[^preprocessing-strategy].  I don't like this approach because there's no way to customize the component rendering for a particular subschema other than specifying a component for _all_ `description` fields... I want hyperlinked schemas to appear visually obvious, and I believe there's no way to hyperlink at all via preprocessing except via the use of a `description` field with a custom `<ReactMarkdown />` component

Since this behavior would dramatically improve the readability of my specification docs, over the next few days I plan to try implementing a proof of concept of this approach in my forked repo. I'd love to get insight from the maintainers here if this sort of use case is worth targeting, and if there's a particular interface or implementation strategy that they'd be most willing to accept. I really want this behavior, and I really don't want to have a lingering fork with this sort of behavior bolted on.

Any advice would be much appreciated 🙏 (and thank you for reading this!)

[^ignoring-equivalent-ids]: For now, I'm ignoring the concern about equivalent IDs (like `schema:my-schema#/$defs/...` vs. unprefixed `#`), but I imagine this to be solvable or ignorable.
[^preprocessing-strategy]: My hacky preprocessing strategy now is that I remove `$ref` fields and instead borrow fields from the referenced schema directly. For the hyperlink, I override with a custom `description` that includes the link.

**Comment by jy95** [comment]
Hello @gnidan ,

For the being time, recursive schemas are not on the [roadmap](https://jy95.github.io/docusaurus-json-schema-plugin/docs/next-steps) yet because I didn't find a clear way to make that obvious to the end-users. So I am curious about your approch (with examples with pictures as I find it helps understanding) ... 

Here is one possible idea I have : 

Thanks to the parsing library used here, unresolved `$ref` like in the case of recursive schema will be present as it in the `schema` object like in [`CreateNodes`](https://github.com/jy95/docusaurus-json-schema-plugin/blob/main/src/theme/JSONSchemaViewer/components/CreateNodes.tsx) . I think a new component `UnresolvedRef` could have its place. Of course, it could be customized as well the `ValueComponent` we are discussing in https://github.com/jy95/docusaurus-json-schema-plugin/pull/224 so users can do whatever they want with it. (In my mind, permalinks could be anything - not only parts on the same page or website if you see my point)

Draft specs about this `UnresolvedRef` could be : 
```
Input 1 : The current JSON Schema at that level 
        (Instead of a standalone "$ref" / "$recursiveRef" / .. input, users will be able to understand the context, 
        not to mention debugging)
Input 2 : The current JSON pointer at that level 
        (I had a proposal to add json pointer  https://github.com/jy95/docusaurus-json-schema-plugin/discussions/120 
        but never got feedbacks on the idea ... )
Output : JSX.Element
```

Let's focus on Input 1 & Output : Input 2 could come later.
(For the default implementation if the custom `UnresolvedRef` isn't here, it could be some kind of placeholder like `Unresolved $ref : xxx` )

What do you think ?

**Reply by gnidan** [reply]
Hm, it's a really interesting idea to use unresolved $refs for this, but I'll need to think about it a bit more. I can probably take that approach to achieve what I want, but I was originally hoping that I could tell each JSONSchemaViewer about my full schema, but ask each instance to show only a portion of that schema. In this, there would be no unresolved $refs (apart from recursive ones), so I wouldn't be able to rely on this fallback behavior.

I was not aware that a recursive $ref simply remains in the schema... that's quite interesting, and would explain some behavior I've observed. I'll have to review `CreateNodes`.

**Reply by jy95** [reply]
@gnidan I created a first proposal around it https://github.com/jy95/docusaurus-json-schema-plugin/pull/225 .
If you are curious about the rendering, you can lookup the [playground](https://deploy-preview-225--delicate-torrone-5d35ee.netlify.app/playground) of that branch

```json
{
	"$id": "https://example.com/tree",
	"$dynamicAnchor": "node",
	"$recursiveAnchor": true,
	"type": "object",
	"properties": {
		"data": true,
		"children": {
			"type": "array",
			"items": {
				"$dynamicRef": "#node",
				"$recursiveRef": "#"
			}
		}
	}
}
```
![image](https://github.com/jy95/docusaurus-json-schema-plugin/assets/9306961/9b8deaf9-3322-4ae7-8566-8ddecfedc966)

Of course, people can customize the rendering like they want thanks to the `UnresolvedRefsComponent ` like I did in my [test](https://github.com/jy95/docusaurus-json-schema-plugin/blob/main/__tests__/JSONSchemaViewer/constructor.test.tsx#L95)

**Reply by gnidan** [reply]
Hey @jy95, sorry for the few days of radio silence... I've been trying to make some headway on my actual schema design work ;)

I've started playing with your new UnresolvedRefsComponent stuff, and I don't think it solves my use case here. I had another thought, though, to achieve what I want: **what if we provide a means for users to specify a custom component mapping for json-pointers?**

I'm imagining something like this:

```tsx
<JSONSchemaViewer
  {...otherOptions}
  viewerOptions={{
    customComponents: {
      "schema:my-schema#/$defs/Foo": (props) => <>/* ... */</>
    }
  }} />
```

I figure this might be a nice way to achieve maximum flexibility without requiring JSONSchemaViewer to know about $refs (since I have learned that you pass that concern over to another package ;).

I do have one concern with this approach: in my custom components, I might want to re-use some or all of the default component rendering logic. I'm imagining these two situations:
- Conditionally skip the custom component and just render the schema as usual
- Render parts of the schema as usual, but modify it partially. Like, perhaps I want all the normal qualifier messages _plus_ extra messaging inside the `<Collapsible />`

Do these make sense? I can start trying to mock this up as I get a sense for my specific requirements.

**Reply by jy95** [reply]
Hello @gnidan , Don't worry even on vacation I've got other things to keep me busy ^^
I'm a bit puzzled about your proposal. Let me explain 3 of my reasons behind : 

1) >  I might want to re-use some or all of the default component rendering logic

In Docusaurus, you have the concept of [Swizzling](https://docusaurus.io/docs/swizzling).
It has been already useful here in the context of https://github.com/jy95/docusaurus-json-schema-plugin/discussions/80 (example of it : https://github.com/jy95/docusaurus-json-schema-plugin/discussions/80#discussioncomment-5896091  ) where some customizations were wanted for `CreateObject` / `CreateArray` but were happy for the other `CreateXXX` components

2) > customComponents: { "schema:my-schema#/$defs/Foo": (props) => <>/* ... */</> }

The hard points are the specs around it and the inherent complexity :

- I would like to integrate json-pointer as id attributes first : https://github.com/jy95/docusaurus-json-schema-plugin/discussions/120#discussioncomment-6532321 before adding this feature
- JSON Pointer have to be properly encoded to avoid issues like "It doesn't work" (people tend to skip the documentation)
- Because some library might simplify the provided schema (for obvious reason of understanding : human hate complexity ^^), JSON Pointers might be not the same ones like the final schema : which is why I think it's important for the user to see the JSON pointer like in https://github.com/jy95/docusaurus-json-schema-plugin/discussions/120#discussioncomment-6532321 without having to delve into the code.
- To determine which `Collapsible` to use (mine or your custom ones), you have to know the current pointer first so again the first point of this list ;)
- We could have the same "reuse" issue like in 1) where Swizzling is the proper way as some parts are internal and may change over time

3) > Conditionally skip the custom component and just render the schema as usual. Render parts of the schema as usual, but modify it partially. Like, perhaps I want all the normal qualifier messages plus extra messaging inside the <Collapsible />

I'd like to avoid creating a mental knot and a technical debt to everyone. 
`DescriptionComponent` / `ValueComponent` / `UnresolvedRefsComponent` have simple goals but they are quite powerful.
You have total freedom and are not dependent on any arbitrary implementation (mine or anyone else's).

As in our case, we depend sooner or later on an internal party, I am more in favour of [Swizzling](https://docusaurus.io/docs/swizzling).
Of course, the component needs to know its current JSON Pointer in order to do its job properly.
I can imagine a new React Context (in a file other than jsvOptions.tsx) to provide this information.

React has a good example of this with "headings" : https://react.dev/learn/passing-data-deeply-with-context#
We could call ours "JSONPointerContext".

**Reply by gnidan** [reply]
Ah sorry for my confusion. I'm starting to understand Swizzling, and what you've said makes sense. I concur that Swizzling + JSON pointer awareness would pretty much solve my use case.

Towards the end of last week I realized the necessity of JSON Pointer awareness, and I started looking at how that might be implemented. I was also thinking that maybe contexts are the right way to go: create a new ContextProvider with the updated JSON pointer value whenever we recurse onto a new schema node.

Want another PR from me? I'm not sure about the display concerns (e.g. an ID inside the HTML), but I should be able to string the JSON pointers logic down the tree :)

**Reply by jy95** [reply]
@gnidan No worries, it took me a while to figure out how it worked too (especially with path aliasing as plugin creator)

If you wish, it would be a pleasure : I'm busy with the learning curve of Flutter and relax time. 
It's important to get some rest sometimes.

About that PR, we will start with something simple : 

1. Implement the new ContextProvider - it could have two properties : 

- `jsonPointer` the. JSON Pointer.  Default value could be the empty string so that easier to concat in nested levels
- `level` the level order in the JSON Schema tree. Default value is 0 (like React nested headings example). At nested level, it will be value of the previous level + 1

![image](https://github.com/jy95/docusaurus-json-schema-plugin/assets/9306961/93bd73ad-b398-4eb6-9f38-b8c5bc502da9)

The level could be useful if we consider later a feature to have [SchemaItem](https://github.com/jy95/docusaurus-json-schema-plugin/blob/main/src/theme/JSONSchemaViewer/components/SchemaItem/SchemaItem.tsx#L66) open by default up to a given level. 

2. On each `CreateEdge` usage ( `CreateItems`, `CreatePrefixItems` , ...), put the context provider and its information. Thanks to `CreateEdge` usages across the code, you know which "context" to generate the proper json pointer (you can see I put some `key` in the past, as I already had some thinking about that)

3. In `SchemaItem`, use the provided information from context. The [id](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id) attribute will be on the `<li className={styles.schemaItem}>`

4. In [JSONSchemaInnerViewer](https://github.com/jy95/docusaurus-json-schema-plugin/blob/main/src/theme/JSONSchemaViewer/index.tsx#L68), add the ContextProvider (so that contexts can work as expected)

5.  Create a new test file with test scenarios : 

You can take the 3 examples I put on https://github.com/jy95/docusaurus-json-schema-plugin/discussions/120#discussioncomment-6532321 and their expecting result `/properties/street_address` , `/prefixItems/3` and `/properties/country~1state` . It should help use track down potential issues with the implementation (such as the encoding rules I learnt in JSON Pointer). With the resulting snapshot and vercel preview builds, we could see the final result of it

Does this proposal make sense ?

**Reply by gnidan** [reply]
From what I remember about how the code is structured, your implementation proposal makes sense to me. Looks like you'll be getting this as a late Christmas present... don't think I'll be on the computer much the next few days :)

Enjoy the season!

**Reply by gnidan** [reply]
Hope all is going well with you over the holidays!

I started picking this up, and I noticed that schema composition and schema conditionals don't use `<CreateEdge />` and instead just jump straight to doing `<CreateNode />`s. I assume the purpose here is to skip the outer wrapping (collapsible) that `<SchemaItem />` provides, but it's a bit unfortunate because it means that there isn't currently a ubiquitous node component that exists at every recursive level.

What do you think we should do about this? I could add the context provider around the `<CreateNodes />` calls, but it seems like there ought to be another component here (either rework `<SchemaItem />` or add something new)

**Reply by jy95** [reply]
@gnidan Like good things, it will have an end soon ^^

For schema composition and schema conditionals, we could add the context provider around the <CreateNodes /> calls for the being time. You don't have to put the ids in the html yet : we need some thinking.

About that another component (or rework of SchemaItem), what would be your idea ? My thoughts : 

- I need the distinction between `<CreateEdge />` and `<CreateNodes />` (e.g. to know when I need the Collapsible and the friendly name generator, ...)
- From a user perspective, short paths would be easier understood and testable :  `/properties/config` instead of `/properties/config/allOf/0` , not to mention if they planned some kind of migration to refactor their schema as well
- `CreateNodes` could put the id in the React frament (instead of the `<>` empty element) but obviously it needs to have the context provider in the various schema conditionals and compositions cases. It will just not be collapsible 
- I am open to discuss about schema composition and schema conditionals rendering refactoring, just first things first. The good point is that as code is specialised, we can tweak and test proposals (for example, having a Collapsible component for both so that we could put id somewhere)

**Reply by gnidan** [reply]
Hm, so I want to keep thinking about this a bit more, but I think that maybe there's some way to separate a component for the full `<Collapsible />` schema container and have something like `<SchemaItem />` be the primary component for rendering a subschema, with HTML DOM ID and `<CreateNodes />` both.

On first glance, I was thinking that some of the behavior from `<SchemaItem />` could be moved to a `<SchemaContainer />` component that takes `children`, but it does look a bit messy.

Anyway, just to respond to one thing you said...

<details>
<summary>Not the main point, but regarding short paths being better.</summary>

> * From a user perspective, short paths would be easier understood and testable :  `/properties/config` instead of `/properties/config/allOf/0`

Sure, but what about this sort of schema:
```yaml
allOf:
  - oneOf:
    - type: object
      properties:
        foo:
          const: "a"
```

Obviously contrived, but you can see how the subschema will need its JSON pointer context to include the `/allOf/0/oneOf/0` bit, so long paths might be unavoidable depending on the schema.

</details>

I'll keep poking around here and see if I get further inspiration.

**Reply by jy95** [reply]
@gnidan Regardless of the component that will use the pointer, we need to specify the context provider. Since code talks more than words sometimes ...

https://github.com/jy95/docusaurus-json-schema-plugin/pull/234 Having a second pair of eyes could help ;) It think it should solve 99% of our present issues.

PS: your example should work with my suggested changes. The fact I don't put the id in the html for `CreateNotes` case yet is to move implementation forward. Putting id on nodes could come later.

**Reply by jy95** [reply]
@gnidan Now that 1.10.0 is out, let's see how it goes for you

**Reply by gnidan** [reply]
I got it working for my purposes last night :D 

Unfortunately I had to do some janky stuff... the JSON pointers themselves weren't actually that useful - since your ref resolver removes the `$ref` fields, a given schema component doesn't know its "canonical" ID+pointer, only the ID+pointer from the viewer root. This turned out to be a bit annoying because I did not want to have long tables of "this pointer in this schema corresponds to this section of the docs", so I ended up just inserting a unique symbol via preprocessing, to contain the ID+pointer at every level. I'm using the changes from 1.10 to check `level` only (see [here](https://github.com/ethdebug/format/pull/55/files#diff-f861e3bad731531271a7b8a6b4ead9f05f46d83e65ea1182d71d9c236aea7a79R21)).

You can see how I transform my schema to insert IDs [here](https://github.com/ethdebug/format/pull/55/files#diff-073d5fe346ba2bfe0f86260d89af9d897c456d1b6c89a7df3f2838400a9d2606R30). Not super nice, but it is passable for me I think. I'm able to define a "schema index" with all the HREFs I need... you can see that [here](https://github.com/ethdebug/format/pull/55/files#diff-79133381edc7334e89d20f0a1a66bfd3926566c49b575a593aa2d23510dc8c0bR54-R100) if you're curious.

Anyway, it works! ([Deployed preview](https://bafybeigreznc4qiowujvgqmfm5n7j52ciuzq5rf3cx6wxgy34tjer7gmlu.on.fleek.co/format/spec/type/#elementary-type-schema))

<details>
<summary>Screenshot of desired behavior working now</summary>

![image](https://github.com/jy95/docusaurus-json-schema-plugin/assets/151065/513cfb28-2f59-4426-a56f-505d3aad9b63)

</details>

Thank you for all your help on this! I'm sure I'll talk to you again soon as I continue to leverage this library :)

**Comment by gnidan** [comment]
Just realized that this discussion is still open. There's probably more to discuss here, but in the end the changes that resulted from this (namely the new context stuff in #234) enable me to achieve what I was looking for with only a small bit of ugly hackery.

I think we can close this until someone else wants something better :)
